[FLINK-18361] Support username and password options for new Elasticse…

KarmaGYZ opened the pull request:
…arch connector

Co-authored-by: zhisheng17 <zhisheng2018@gmail.com>

## What is the purpose of the change

 Support username and password options for new Elasticsearch connector.

Note that this PR tries to align the behavior with #11822 , so I add @zhisheng17 as co-auther.

## Brief change log

- Add `username` and `password` option to new ES DynamicTableSink.
- If both `username` and `password` configured, add a `UsernamePasswordCredentials` to `RestClientBuilder`.

## Verifying this change

This change added tests and can be verified as follows:

- Elasticsearch7DynamicSinkTest#testAuthConfig
- Elasticsearch7DynamicSinkFactoryTest#validateWrongCredential
- Elasticsearch6DynamicSinkTest#testAuthConfig
- Elasticsearch6DynamicSinkFactoryTest#validateWrongCredential

As the authentication logic we used is documented in [ES official doc](https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/_basic_authentication.html). I think Flink does not need to check it in an e2e test or integration test. However, I'll manually test it and record in this PR

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

KarmaGYZ commented:
I have manually tested it with ES 6 and 7.

KarmaGYZ commented:
Would you like to take a look @wuchong ?

wuchong commented:
Thanks for ping me. I will have a look later today.

wuchong commented on docs/dev/table/connectors/elasticsearch.md:
```diff
@@ -126,6 +126,20 @@ Connector Options
+      <td>optional</td>
+      <td style="word-wrap: break-word;">(none)</td>
+      <td>String</td>
+      <td>Username used to connect to Elasticsearch instance.</td>
```
What do you think about adding more explanation about ES authentication here? I guess not all the users know the auth feature of Elasticsearch. For example:

```
Elasticsearch doesn't pre-bundled security feature, but you can enable it by following the [guideline](https://www.elastic.co/guide/en/elasticsearch/reference/7.x/configuring-security.html) to secure an Elasticsearch cluster.
```

KarmaGYZ commented:
Thanks for the review, @wuchong . I've updated the documentation and rebased it on the latest master.

Regarding the test, I also agree to verify it in an e2e test and have another ticket to track it.

wuchong commented:
@flinkbot run azure

wuchong commented:
`[FAIL] 'SQL Client end-to-end test (Blink planner) Elasticsearch (v7.5.1)' failed after 0 minutes and 27 seconds! Test exited with exit code 1`

The ES e2e test is failed.

wuchong commented:
`[FAIL] 'SQL Client end-to-end test (Blink planner) Elasticsearch (v7.5.1)' failed after 0 minutes and 27 seconds! Test exited with exit code 1`

The ES e2e test is failed: https://dev.azure.com/apache-flink/apache-flink/_build/results?buildId=4368&view=logs&j=c88eea3b-64a0-564d-0031-9fdcd7b8abee&t=ff888d9b-cd34-53cc-d90f-3e446d355529

KarmaGYZ commented:
@wuchong It seems that the failure is caused by network error during downloading the elasticsearch.

KarmaGYZ commented:
@flinkbot run azure

wuchong commented:
Merging...
